[WellSQL Migration] Migrate `WCGateways` to Room

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: [AINFRA-551](https://linear.app/a8c/issue/AINFRA-551)
Relates To: [AINFRA-1190](https://linear.app/a8c/issue/AINFRA-1190)

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCGateways` from [WellSQL](https://github.com/wordpress-mobile/wellsql) to [Room](https://developer.android.com/training/data-storage/room) database.

---

### Reviewers

In addition to having @wzieba as the main "code" reviewer on this change, I added `@woocommerce/android-developers` to randomly assign a product engineer as an "extra" reviewer (@JorgeMucientes), primarily because:

1.  This change is payment related and thus might require a bit more attention. ⚠️
2. Although I was able to identify a couple of testing instructions, I am sure I am missing core payment flows that could/should be tested prior to merging this. Thus, I would very much appreciate if a product engineer would help us identify and test such (missing) flows. 🙏 

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

#### `Orders`

1. Uninstall, build and install the `trunk` version of the app (`WCGateways`).
3. As you open the app for the first time (after fresh install), while on the `My store` tab, open `App Inspection` and verify that you see the `WCGateways` table (within the `wp-fluxc` database), which should be empty.
4. Navigate on the `Orders` tab and notice that `WCGateways` table is being populated with at least a few gateway entries, like `cod` and `stripe`.
5. Now build and install the `migrate/wc-gateways-to-room` version of the app (`GatewayEntity`).
6. As you open the app for the first time (after migration), while on the `My store` tab, open `App Inspection` and verify that you no longer see the `WCGateways` table (within the `wp-fluxc` database), but you do see the `GatewayEntity` table (within the `wc-android-database` database), which should be empty.
7. Navigate on the `Orders` tab and notice that `GatewayEntity` table is being populated with at least a few gateway entries, like `cod` and `stripe`. (thanks to 4d065849df76ceef02bed90fac83c8a549f93eba)

#### `Menu` + `Payments`

1. From the previous set of testing instructions, while you have `App Inspection` open and browsing that `GatewayEntity` table, export the table values using the `Export as file...` option, let's name this file `1-GatewayEntity-Orders.csv`:

<img width="740" height="308" alt="image" src="https://github.com/user-attachments/assets/fe0ae1dd-c27e-438d-9fc1-82b214ef38f7" />

2. Navigate on the `Menu` tab, then click on the `Payments` option, and while on the `Payments` screen toggle to enable the `Pay in Person` switch.
3. Export the table values using the `Export as file...` option again, name this file `2a-GatewayEntity-Menu-Payments-PayInPerson-On.csv`.
4. Compare the two files, `1-GatewayEntity-Orders.csv` and `2a-GatewayEntity-Menu-Payments-PayInPerson-On.csv` and notice that the gateway entry for `cod` indeed gets updated from `"enabled":false` to `"enabled":true` for the `data` column.
5. While on the `Payments` screen, toggle to disable the `Pay in Person` switch.
6. Export the table values using the `Export as file...` option again, name this file `2b-GatewayEntity-Menu-Payments-PayInPerson-Off.csv`.
8. Compare the two files, `2a-GatewayEntity-Menu-Payments-PayInPerson-On.csv` and `2b-GatewayEntity-Menu-Payments-PayInPerson-Off.csv` and notice that the gateway entry for `cod` indeed gets updated from `"enabled":true` to `"enabled":false` for the `data` column.

#### `Orders` + `Refund`

1. Go back to `Orders` screen and find a `Completed` order.
2. Find and click the `Issue refund` button.
3. Complete the refund flow and verify everything is working as expected.

---

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

<img width="1918" height="890" alt="image" src="https://github.com/user-attachments/assets/0635caa3-34db-41f2-a697-83d1555fce8f" />

---

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Test: Simplify wc global attribute store test

1. No need to setup 'Robolectric'.
2. No need to init mocks on setup. Mocked fields are such as good.
3. No need to spy on the mapper for that one specific test.

ParaskP7 committed:
Cleanup: Remove unused fetch gateway related store/rest code

ParaskP7 committed:
Refactor: Rename update payment gateway to update gateway

Every other store/rest function doesn't include the 'payment' keyword.
Keeping things consistent here is better, in order to avoid having to
think twice whether 'gateway' and 'payments gateway' refer to a
different API, which they don't. Everything points to:

WOOCOMMERCE.payment_gateways

ParaskP7 committed:
Test: Add couple of missing tests to wc gateway store test

ParaskP7 committed:
Refactor: Move gson out of wc gateway sql utils and into gateway mapper

ParaskP7 committed:
Refactor: Reformat wc gateway store (test)

ParaskP7 committed:
Test: Refactor gateway fixtures

ParaskP7 committed:
Refactor: Make getAllGateways a suspend function

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Test: Move gateway test fixtures and store test to fluxc-plugin

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Migrate: 'WCGateways' from wellsql to room

FYI: This change includes but is not limited to the:
- Entity Introduction
- DAO Introduction (With Test)
- WCAndroidDatabase Migration
- Code Adjustments (Mapper, Store)
- Test Adjustments (Mapper, Store)

------------------------------------------------------------------------

Also, as you'll compare '59.json' to '60.json' you'll notice quite a lot
of changes, which are unrelated to the introduction of this new
'GatewayEntity'. This is (most probably) because as part of this
a55f6a499aaba43fd5fa35fa348bbcd35186a891 change, 'androidx-room' got
updated from '2.6.1' to '2.7.2' .

ParaskP7 committed:
Db: Drop 'WCGateways' table from 'wp-fluxc'

ParaskP7 committed:
Test: Add gateway mapper test

ParaskP7 committed:
Fix: Fetch order list after migrating gateway from wellsql to room

FYI: This is done because after migrating 'WCGateways' from WellSQL to
'GatewayEntity' in Room, when navigating to 'Orders' and unless manually
refreshing them, the 'GatewayEntity' table remains empty (while
prior to the migration it was populated with at least a few gateway
entries).

Related: https://linear.app/a8c/issue/AINFRA-1190

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/267">23.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Analysis: Resolve/Suppress all detekt warnings

1. Resolve "UnitTestNamingRule":
   - GatewayMapperTest.kt
   - WCGatewayStoreTest.kt
   - GatewayRestClientTest.kt
2. Resolve "ImportOrdering":
   - WCGatewayStoreTest.kt
3. Suppress "ReturnCount":
   - ShouldUpdateOrdersList.kt

PS: Also, as part of this change, those 8 "UnitTestNamingRule" related
baseline suppression got resolved/removed as well, them no longer being
relevant since they are now fixed.

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/GatewaysDao.kt:
```diff
@@ -0,0 +1,49 @@
+        AND gatewayId = :gatewayId
+        LIMIT 1
+        """
+    )
```
```suggestion
    @Query(
        """
        SELECT * FROM GatewayEntity
        WHERE siteId = :siteId
        AND gatewayId = :gatewayId
        """
    )
```

`siteId` and `gatewayId` are already guaranteed to be unique, so this query will always return 1 result if there's any match so `LIMIT 1` doesn't provide any benefits - I suggest we remove it.

ParaskP7 replied:
FYI: This is now done: dc01601f92a741f5e251454ceef07cb494601794 @wzieba

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/GatewaysDao.kt:
```diff
@@ -0,0 +1,49 @@
+    protected abstract suspend fun deleteForSite(siteId: LocalId)
+
+    @Upsert
+    abstract suspend fun upsertGateways(gateways: List<GatewayEntity>)
```
```suggestion
    protected abstract suspend fun upsertGateways(gateways: List<GatewayEntity>)
```

this method is used only in this class

ParaskP7 replied:
FYI: This is now done: 23ad358eaec794b610e5e4fe39cba0e2022fc783 @wzieba

wzieba commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/store/WCGlobalAttributeStoreTest.kt:
```diff
@@ -1,50 +1,35 @@
-    private lateinit var mapper: WCGlobalAttributeMapper
+    private val restClient = mock<ProductAttributeRestClient>()
+    private val globalAttributesDao = mock<GlobalAttributesDao>()
+    private val mapper = mock<WCGlobalAttributeMapper>()
```
I know you didn't introduce this, but please let's not mock mappers

```suggestion
    private val mapper = WCGlobalAttributeMapper()
```

ParaskP7 replied:
FYI: This is now done: 51df2f51961259007249a4eb2f8a490370f83bf7 @wzieba

wzieba commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/store/WCGlobalAttributeStoreTest.kt:
```diff
@@ -62,8 +47,6 @@ class WCGlobalAttributeStoreTest {
 
     @Test
     fun `fetch attributes should call mapper once`() = test {
```
I'd vote to remove this test. There's no point in checking that sut "should call mapper" - it's an implementation detail, the test doesn't focus on output at all.

ParaskP7 replied:
FYI: This is now done: 29dd7e2c9656ce54feeaf996ef7f7277842662d6 @wzieba

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/ShouldUpdateOrdersList.kt:
```diff
@@ -13,13 +13,20 @@ class ShouldUpdateOrdersList @Inject constructor(
             appPrefs.orderSummaryMigrated = true
             return true
         }
+        // 26-08-2025: Consider removing this in the future
+        // AINFRA-1190
+        if (!appPrefs.gatewayMigrated) {
+            appPrefs.gatewayMigrated = true
+            return true
+        }
```
In the commit message I see

> FYI: This is done because after migrating 'WCGateways' from WellSQL to
'GatewayEntity' in Room, when navigating to 'Orders' and unless manually
refreshing them, the 'GatewayEntity' table remains empty (while
prior to the migration it was populated with at least a few gateway
entries).

What's the consequence of the empty `GatewayEntity` table when user navigates to `Orders`?

In case of "order summaries" it was an empty view (no orders visible), so that's why we introduced this one-off force fetch.

ParaskP7 replied:
👋 @wzieba !

> What's the consequence of the empty GatewayEntity table when user navigates to Orders?

Great question, to be honest, I am not really sure. Thus, I wanted someone from product engineering to help me figure this out. 🤔

> [Although I'm no expert in the payments features from the app.](https://github.com/woocommerce/woocommerce-android/pull/14520#pullrequestreview-3167715765)

@JorgeMucientes, which team, `Kiwi` or `Peacock` is best suites to help with such payments related change, just so we build some more confidence that what we're doing here, specifically with this fetching of gateways, is correct? 🙏

> In case of "order summaries" it was an empty view (no orders visible), so that's why we introduced this one-off force fetch.

Yea, so the main reason I introduce this is because I noticed that during a fresh install, when the list of orders gets fetched, the list of gateways get fetched as well.

So, I thought let's use the same tricks, which would be quite invisible for the user and re-fetch the orders during this migration too, just so that we achieve the same end result. 🤔

ParaskP7 replied:
FYI: See: [comment](https://github.com/woocommerce/woocommerce-android/pull/14520#issuecomment-3245335184) @wzieba

JorgeMucientes commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/gateways/GatewayRestClient.kt:
```diff
@@ -10,21 +10,7 @@ import org.wordpress.android.fluxc.utils.toWooPayload
 import javax.inject.Inject
 
 class GatewayRestClient @Inject constructor(private val wooNetwork: WooNetwork) {
-    suspend fun fetchGateway(
-        site: SiteModel,
-        gatewayId: String
-    ): WooPayload<GatewayResponse> {
```
Removed because unused right?

ParaskP7 replied:
Yes, 100% @JorgeMucientes , this is removed because it is unused. 👍

JorgeMucientes reviewed (approved):
Great job @ParaskP7 thanks for addressing this migration 🙏🏼 🏅 

I didn't find any issues  and all the test scenarios you shared worked as expected. I also tested that updating the payment gateways from `wc-admin` will also end up refreshing the values on mobile. For example enabling/disabling "cod": 

<img width="915" height="359" alt="Screenshot 2025-08-29 at 09 05 32" src="https://github.com/user-attachments/assets/7a733191-83ea-4226-90a2-15d75f6ac382" />

Aside of that I can't think of any other scenarios to test. Although I'm no expert in the payments features from the app. 

LGTM :shipit:

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-gateways-to-room

# Conflicts:
#	libs/fluxc-tests/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/GatewayRestClientTest.kt

ParaskP7 commented:
Awesome @JorgeMucientes , thanks for reviewing and testing this change, much appreciated! 🙇 ❤️ 🚀 

> I didn't find any issues and all the test scenarios you shared worked as expected. I also tested that updating the payment gateways from wc-admin will also end up refreshing the values on mobile. For example enabling/disabling "cod":

🙏

> Aside of that I can't think of any other scenarios to test. Although I'm no expert in the payments features from the app.

As per [this](https://github.com/woocommerce/woocommerce-android/pull/14520#discussion_r2313507499) comment of mine, maybe we should bring an additional product engineer to get another look at that, someone with some more payments expertise, wdyt? 🤔

JorgeMucientes commented:
> As per https://github.com/woocommerce/woocommerce-android/pull/14520#discussion_r2313507499 comment of mine, maybe we should bring an additional product engineer to get another look at that, someone with some more payments expertise, wdyt? 🤔

My take. 

I think the scenarios you shared in the PR instructions are fairly exhaustive. I don't think a full review from another engineer is needed, but I guess it wouldn't harm to have @kidinov, @malinajirka or @samiuelson  check if the PR instructions missed any relevant flow that could be impacted by this DB migration. And if that were the case share some instructions on how to test it or test it themselves 🙏🏼

ParaskP7 commented:
Many thanks @JorgeMucientes and also [this](https://github.com/woocommerce/woocommerce-android/pull/14520#discussion_r2303833115) folks (@kidinov, @malinajirka or @samiuelson), your input much appreciated: 🙏

> What's the consequence of the empty GatewayEntity table when user navigates to Orders?

kidinov commented:
@ParaskP7 👋

> Many thanks @JorgeMucientes and also https://github.com/woocommerce/woocommerce-android/pull/14520#discussion_r2303833115 folks (@kidinov, @malinajirka or @samiuelson), your input much appreciated: 🙏
> What's the consequence of the empty GatewayEntity table when user navigates to Orders?

Tbh I think none of us ever worked with this part of the app, but I looked up from my side what I could understand. 

I think the only side effect I see is  In `ShouldShowCreateTestOrderScreen.kt (line 21): orderListRepository.getAllPaymentGateways(site).isNotEmpty()`:
  - This check will return false when gateways table is empty
  - The "Create Test Order" empty view won't show even if it should
  - Users might see a generic empty state instead of the test order creation prompt

Was this reproducible and noticiable during your testing?

ParaskP7 commented:
👋 @kidinov and thanks for helping! 🙇 

> Tbh I think none of us ever worked with this part of the app, but I looked up from my side what I could understand.

👍 and thanks, much appreciated. 🙏

> I think the only side effect I see is In ShouldShowCreateTestOrderScreen.kt (line 21): orderListRepository.getAllPaymentGateways(site).isNotEmpty():

👍 

> Was this reproducible and noticiable during your testing?

I just created [this](https://reliable-reef-galapagos.jurassic.ninja/wp-admin/index.php) Jurassic.Ninja test site, added a test product, and was able to reproduce this:

- Without this 4d065849df76ceef02bed90fac83c8a549f93eba extra logic to fetch the order list after migrating I got this generic `Learn more` screen:

<img width="162" height="360" alt="image" src="https://github.com/user-attachments/assets/d45864e0-9820-4e44-a619-5ea643a0b947" />

- With this 4d065849df76ceef02bed90fac83c8a549f93eba extra logic to fetch the order list after migrating I got the expected `Try a test order` screen:

<img width="162" height="360" alt="image" src="https://github.com/user-attachments/assets/55ca1a61-08bf-453c-9c65-304c8a875b6e" />
 
---

Your thoughts? 💭

kidinov commented:
@ParaskP7 

So with this commit it works the same as it used to be? If it is then I guess we are good, right?

ParaskP7 commented:
> So with this commit it works the same as it used to be? If it is then I guess we are good, right?

Yeap, exactly @kidinov , I guess so... 😊 

@wzieba your thoughts? 💭 

PS: I'll then proceed with your (other) existing code-review comments and apply your suggested changes. 🙏

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-gateways-to-room

ParaskP7 committed:
Db: Remove unnecessary limit 1 from get gateway on gateways dao

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14520#discussion_r2303799665

ParaskP7 committed:
Refactor: Make upsert gateways function protected on gateways dao

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14520#discussion_r2303803405

ParaskP7 committed:
Test: Unmock mapper on wc global attribute store test

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14520#discussion_r2303819771

ParaskP7 committed:
Test: Remove unimpactful test from wc global attribute store test

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14520#discussion_r2303824869

wzieba commented:
Thanks for sharing the flow @kidinov and @ParaskP7 for preparing the example 🙏 ! I'm onboard with having the additional force-fetch to keep the previous behavior - it _is_ needed.

wzieba reviewed (approved):
Thank you for addressing comments 🙏 ! LGTM :shipit:

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-gateways-to-room

ParaskP7 commented:
Awesome, thank you all @wzieba @kidinov and @JorgeMucientes , as soon as CI goes 🟢 I'll be merging this! 🙇 ❤️ 🚀
